cpp/src/prims/transform_e.cuh

```diff
@@ -0,0 +1,342 @@
+        edge_partition_src_input_device_view_t(edge_src_value_input, i);
+      edge_partition_dst_value_input = edge_partition_dst_input_device_view_t(edge_dst_value_input);
+    }
+    auto edge_partition_e_value_input = edge_partition_e_input_device_view_t(edge_value_input, i);
+    auto edge_partition_e_value_output =
+      edge_partition_e_output_device_view_t(edge_value_output, i);
+
+    thrust::for_each(
+      handle.get_thrust_policy(),
+      edge_first + edge_partition_offsets[i],
+      edge_first + edge_partition_offsets[i + 1],
+      [e_op,
+       edge_partition,
+       edge_partition_src_value_input,
+       edge_partition_dst_value_input,
+       edge_partition_e_value_input,
+       edge_partition_e_value_output] __device__(thrust::tuple<vertex_t, vertex_t> edge) {
+        auto major = thrust::get<0>(edge);
+        auto minor = thrust::get<1>(edge);
+
+        vertex_t major_offset{};
+        vertex_t major_idx{};
+        auto major_hypersparse_first = edge_partition.major_hypersparse_first();
+        if (major_hypersparse_first) {
+          if (major < *major_hypersparse_first) {
+            major_offset = edge_partition.major_offset_from_major_nocheck(major);
+            major_idx    = major_offset;
+          } else {
+            auto major_hypersparse_idx =
+              edge_partition.major_hypersparse_idx_from_major_nocheck(major);
+            assert(major_hypersparse_idx);
+            major_idx = edge_partition.major_offset_from_major_nocheck(*major_hypersparse_first) +
+                        *major_hypersparse_idx;
+          }
+        } else {
+          major_offset = edge_partition.major_offset_from_major_nocheck(major);
+          major_idx    = major_offset;
+        }
+
+        auto minor_offset = edge_partition.minor_offset_from_minor_nocheck(minor);
+
+        vertex_t const* indices{nullptr};
+        edge_t edge_offset{};
+        edge_t local_degree{};
+        thrust::tie(indices, edge_offset, local_degree) = edge_partition.local_edges(major_idx);
+        auto lower_it = thrust::lower_bound(thrust::seq, indices, indices + local_degree, minor);
+        auto upper_it = thrust::upper_bound(thrust::seq, indices, indices + local_degree, minor);
+
+        auto src        = GraphViewType::is_storage_transposed ? minor : major;
+        auto dst        = GraphViewType::is_storage_transposed ? major : minor;
+        auto src_offset = GraphViewType::is_storage_transposed ? minor_offset : major_offset;
+        auto dst_offset = GraphViewType::is_storage_transposed ? major_offset : minor_offset;
+
+        for (auto it = lower_it; it != upper_it; ++it) {
+          assert(*it == minor);
+          auto e_op_result =
+            e_op(src,
+                 dst,
+                 edge_partition_src_value_input.get(src_offset),
+                 edge_partition_dst_value_input.get(dst_offset),
+                 edge_partition_e_value_input.get(edge_offset + thrust::distance(indices, it)));
```

Comment:
General question:
It would results in a multiple iteration only if there are multiple edges between (major, minor)?

Reply:
Yes, correct.